### PR TITLE
feat: add power_analysis() with sigmoid fitting strategy (#820)

### PR DESCRIPTION
**Pull request by genrichez**
## Summary

Implements efficient power curve estimation for geo-experiment design, as requested in #820.

### `power_analysis()` function
Computes detection probability as a function of true effect size, using the learned null distribution from a completed `PlaceboInTime` check. Two strategies:

- **`grid`** — evaluate at every user-supplied effect size (brute-force Monte Carlo)
- **`sigmoid`** — evaluate at ~5 points, fit a two-parameter logistic, extract MDE analytically

### New public API
- `causalpy.checks.power_analysis(pit_result, ...)` — main entry point
- `PowerCurveResult` — dataclass with `.plot()` method
- `LogisticFit` — fitted logistic with `.predict()` and `.mde()` methods

## Usage

~~~python
import causalpy as cp

# After running a PlaceboInTime check:
# result = pipeline.run()
# pit_check = result.sensitivity_results[0]

# Grid strategy (brute-force)
curve = cp.checks.power_analysis(
    pit_check,
    effect_sizes=[0, 50, 100, 150, 200, 300],
    n_simulations=200,
    strategy="grid",
)
curve.plot()

# Sigmoid strategy (efficient)
curve = cp.checks.power_analysis(
    pit_check,
    effect_sizes=[0, 300],  # range bounds
    n_simulations=200,
    strategy="sigmoid",
    n_evaluation_points=5,
)
print(f"MDE at 80% power: {curve.mde:.2f}")
curve.plot()
~~~

## Validation

Compared 5-point sigmoid vs 20-point grid (ground truth):

| Metric | Value |
|--------|-------|
| Max absolute error | 1.4% |
| Mean absolute error | 0.6% |
| Speedup (5 pts vs 20 pts) | 3.8x |
| Computation reduction | 74% |

Robust across different null parameter regimes (null_std 20–200, rope 10–50).

## Tests

42 tests total (27 pytest + 15 edge case/business logic), covering:
- Logistic function properties and LogisticFit predict/MDE extraction
- Detection rate simulation (monotonicity, bounds)
- Grid strategy (structure, reproducibility, increasing detection)
- Sigmoid strategy (fitting, MDE, smooth curve, default ranges)
- Error handling (missing metadata, invalid strategy)
- Plotting (grid, sigmoid, custom axes)
- Edge cases: all-zero/all-saturated detection, very few null samples, negative effects, low n_simulations, heterogeneous fold SDs, degenerate ROPE (rope=0), biased null, single effect size, very large effects (1e9, no overflow), partial transition sigmoid, threshold sensitivity

## Design Notes

- The function operates on a completed `PlaceboInTime` `CheckResult`, leveraging the learned null distribution (`null_samples`, `fold_sds`) to simulate detection at hypothetical effect sizes using the same ROPE-based Bayesian decision rule.
- Detection is **one-sided** (`P(posterior > rope) >= threshold`), consistent with how `PlaceboInTime` assurance works. Users wanting two-sided power should use `|effect_size|`.
- The sigmoid model is `P(detect | x) = 1 / (1 + exp(-k*(x - x0)))`. MDE is extracted analytically: `x0 + (1/k) * ln(tau / (1 - tau))`.
- When sigmoid fitting fails (e.g., flat data), a warning is emitted and raw evaluation points are returned without a fitted curve.

Closes #820

### CODE REVIEW

File: causalpy/tests/test_power_analysis.py [hunk 1]
```diff
@@ -0,0 +1,600 @@
...
+    LogisticFit,
+    PowerCurveResult,
+    _logistic,
+    _simulate_detection_rate,
+    power_analysis,
+)
+
+# ---------------------------------------------------------------------------
+# Fixtures
+# ---------------------------------------------------------------------------
```
Comment (juanitorduz) [inline causalpy/tests/test_power_analysis.py]: Can we please remove these descriptions? they are redundant :)

File: causalpy/tests/test_power_analysis.py [hunk 2]
```diff
@@ -0,0 +1,600 @@
...
+            "threshold": threshold,
+            "actual_cumulative_mean": 150.0,
+            "p_effect_outside_null": 0.97,
+        },
+    )
+
+
+# ---------------------------------------------------------------------------
+# Tests for _logistic helper
+# ---------------------------------------------------------------------------
```
Comment (juanitorduz) [inline causalpy/tests/test_power_analysis.py]: and this one

File: causalpy/tests/test_power_analysis.py [hunk 3]
```diff
@@ -0,0 +1,600 @@
...
+        """Higher k should produce steeper transition."""
+        x = np.array([5.5])
+        y_shallow = _logistic(x, k=1.0, x0=5.0)
+        y_steep = _logistic(x, k=10.0, x0=5.0)
+        assert y_steep > y_shallow
+
+
+# ---------------------------------------------------------------------------
+# Tests for LogisticFit
+# ---------------------------------------------------------------------------
```
Comment (juanitorduz) [inline causalpy/tests/test_power_analysis.py]: and this one

File: causalpy/tests/test_power_analysis.py [hunk 4]
```diff
@@ -0,0 +1,600 @@
...
+        fit = LogisticFit(k=0.05, x0=100.0)
+        with pytest.raises(ValueError, match="power_threshold must be in"):
+            fit.mde(0.0)
+        with pytest.raises(ValueError, match="power_threshold must be in"):
+            fit.mde(1.0)
+
+
+# ---------------------------------------------------------------------------
+# Tests for _simulate_detection_rate
+# ---------------------------------------------------------------------------
```
Comment (juanitorduz) [inline causalpy/tests/test_power_analysis.py]: and this one

File: causalpy/tests/test_power_analysis.py [hunk 5]
```diff
@@ -0,0 +1,600 @@
...
+
+        # Not strictly monotonic due to MC noise, but overall trend
+        assert rates[-1] > rates[0]
+        assert rates[-1] > 0.70
+
+
+# ---------------------------------------------------------------------------
+# Tests for power_analysis (grid strategy)
+# ---------------------------------------------------------------------------
+
```
Comment (juanitorduz) [inline causalpy/tests/test_power_analysis.py]: and this one

File: causalpy/tests/test_power_analysis.py [hunk 6]
```diff
@@ -0,0 +1,600 @@
...
+            n_simulations=100,
+            strategy="grid",
+            random_seed=42,
+        )
+        np.testing.assert_array_equal(r1.detection_rates, r2.detection_rates)
+
+
+# ---------------------------------------------------------------------------
+# Tests for power_analysis (sigmoid strategy)
+# ---------------------------------------------------------------------------
```
Comment (juanitorduz) [inline causalpy/tests/test_power_analysis.py]: and this one

**Review by daimon-pymclabs — COMMENTED** [review]
Thanks for this — the sigmoid idea is well-motivated and the module is clean, well-documented, and comes with a genuinely thorough test suite (nice that the tests are pure-numpy against a hand-built `CheckResult`, so CI stays fast with no MCMC). I verified the metadata contract: `PlaceboInTime.run()` does emit `null_samples`, `fold_sds`, `rope_half_width`, and `threshold`, so the function wires up correctly against real output. A few things worth discussing before merge, one design-level and a handful of code-level.

## Design: is this the API we want?

The biggest question is scope alignment. Issue #820 framed this as adding a `strategy="sigmoid"` parameter to the existing `SyntheticControl.power_analysis()` method introduced in #819, and #819 is still open. This PR instead ships a **standalone free function** `cp.checks.power_analysis(pit_result, ...)` that operates on a completed `PlaceboInTime` result rather than a `SyntheticControl` experiment. So we'd end up with two parallel power-analysis surfaces that don't share code or signature:

- `SyntheticControl.power_analysis(...)` (method, from #819)
- `causalpy.checks.power_analysis(pit_result, ...)` (function, this PR)

That's worth reconciling explicitly. Options I see: (a) land the sigmoid strategy inside the #819 method as the issue proposed; (b) keep the free function but factor the shared sigmoid-fitting + `LogisticFit` + `PowerCurveResult` into one place both call; (c) decide the two are genuinely different tools (SC dress-rehearsal vs PlaceboInTime learned-null) and document why. Right now #820's acceptance criteria aren't literally met, so I'd want @juanitorduz to confirm which direction is intended before investing more in either.

## Statistical: null-component sampling (shared with `_compute_assurance`)

The detection simulation draws the null component as `null_samples[i % len(null_samples)]` (deterministic prefix), so with the typical `n_simulations=200` against ~4000 `theta_new` draws you only ever use the first 200 samples. Those draws come from `posterior_predictive["theta_new"].stack(sample=("chain","draw"))`, which orders chain-slowest — so the first 200 are all from chain 0. `random_seed` doesn't change which null draws are used (it only affects the `sigma` pick and the normal noise). This is a real (mild) bias and it's **shared with the existing `PlaceboInTime._compute_assurance`**, which uses the identical `i % len` idiom. Suggest fixing it centrally with `rng.choice(null_samples, size=n_simulations, replace=True)` and reusing that in both places. See inline note.

## Correctness note (in the PR's favor)

The analytic MDE in code is `x0 + (1/k)·ln(tau/(1-tau))`, which is the correct inversion (MDE > x0 for tau>0.5). Note the issue text wrote `ln((1-tau)/tau)`, which has the sign flipped — the **code is right, the issue was wrong**, good catch by whoever implemented it.

## Smaller items (details inline)

- Decision rule is re-implemented inline instead of reusing `PlaceboInTime.bayesian_rope_decision`; they can drift.
- `except RuntimeError` around `curve_fit` is too narrow (scipy raises `ValueError`/`OptimizeWarning` on degenerate data too), and `LogisticFit.mde` divides by `k` with no guard for `k → 0`.
- Docstring defaults for `effect_sizes` don't match the code (`4*tau` vs `max(4*tau, 3*rope)`).
- The `fold_results` → `fr.fold_sd` fallback is effectively dead against real output (whenever `null_samples` exists, `fold_sds` does too), only reachable via hand-built metadata.
- No end-to-end smoke test running an actual tiny `PlaceboInTime.run()` into `power_analysis` — all tests use a synthetic `CheckResult`, so metadata-shape drift wouldn't be caught.

None of the code-level items are blockers; the design question is the one I'd resolve first.

File: causalpy/checks/power_analysis.py [hunk 7]
```diff
@@ -0,0 +1,544 @@
...
+
+    Returns
+    -------
+    float
+        Estimated detection probability (fraction of "positive" decisions).
+    """
+    detections = 0
+    for i in range(n_simulations):
+        # Draw null component (structural noise)
+        null_component = float(null_samples[i % len(null_samples)])
```
Comment (daimon-pymclabs) [inline causalpy/checks/power_analysis.py]: **Null draw is a deterministic prefix, not seeded.** `null_samples[i % len(null_samples)]` walks the first `n_simulations` entries in order. With `n_simulations=200` and ~4000 `theta_new` draws, only the first 200 are ever used, and because `theta_new` is `stack(sample=("chain","draw"))` (chain-slowest) those 200 all come from chain 0. `random_seed` never touches this draw. Prefer `rng.choice(null_samples, size=n_simulations, replace=True)`.

This exact idiom also lives in `PlaceboInTime._compute_assurance`, so ideally fix once and share (a small module-level `_draw_null_components(rng, null_samples, n)` helper both call) rather than duplicating the biased version here.

File: causalpy/checks/power_analysis.py [hunk 8]
```diff
@@ -0,0 +1,544 @@
...
+        # True effect = null noise + injected effect
+        true_effect = null_component + effect_size
+        # Simulate estimation uncertainty
+        sigma = float(rng.choice(fold_sds))
+        simulated_posterior = rng.normal(
+            loc=true_effect, scale=sigma, size=n_posterior_samples
+        )
+        # Apply ROPE decision
+        prob_positive = float((simulated_posterior > rope_half_width).mean())
```
Comment (daimon-pymclabs) [inline causalpy/checks/power_analysis.py:301]: The ROPE decision is re-implemented inline (`prob_positive >= threshold`) instead of calling the existing `PlaceboInTime.bayesian_rope_decision(...) == "positive"`. It's numerically equivalent to the one-sided `"positive"` branch today, but if the shared decision rule ever changes (e.g. ROPE semantics), this copy silently drifts. Reusing the static method keeps power analysis and assurance consistent by construction.

File: causalpy/checks/power_analysis.py [hunk 9]
```diff
@@ -0,0 +1,544 @@
...
+            "one fold."
+        )
+
+    null_samples = np.asarray(meta["null_samples"]).ravel()
+    fold_sds_raw = meta.get("fold_sds")
+    if fold_sds_raw is None:
+        fold_results = meta.get("fold_results", [])
+        if not fold_results:
+            raise ValueError("pit_result has no fold_sds or fold_results in metadata.")
+        fold_sds_raw = [fr.fold_sd for fr in fold_results]
```
Comment (daimon-pymclabs) [inline causalpy/checks/power_analysis.py]: This `fold_results` → `fr.fold_sd` fallback is effectively unreachable against real `PlaceboInTime` output: whenever `null_samples` is present (required a few lines up), the run also stored `fold_sds`. The 0-folds-completed path that lacks `fold_sds` also lacks `null_samples`, so it fails the earlier check first. It's only exercised by hand-built metadata in the tests. Not harmful, but worth a comment noting it's a defensive path, or drop it.

File: causalpy/checks/power_analysis.py [hunk 10]
```diff
@@ -0,0 +1,544 @@
...
+
+    threshold = float(meta.get("threshold", 0.95))
+
+    # Compute null scale for default ranges
+    tau = float(null_samples.std())
+
+    # Set up effect sizes based on strategy
+    if strategy == "grid":
+        if effect_sizes is None:
+            effect_sizes_arr = np.linspace(0, max(4.0 * tau, 3.0 * rope_half_width), 8)
```
Comment (daimon-pymclabs) [inline causalpy/checks/power_analysis.py]: Docstring drift: the docstring says the default is `np.linspace(0, 4 * tau, 8)` (grid) and `[0, 4 * tau]` (sigmoid), but the code uses `max(4.0 * tau, 3.0 * rope_half_width)` as the upper bound in both. Sync the docstring so users know the ROPE term participates in the default range.

File: causalpy/checks/power_analysis.py [hunk 11]
```diff
@@ -0,0 +1,544 @@
...
+
+            # Warn if MDE is outside the evaluated range
+            if mde < effect_sizes_arr[0] or mde > effect_sizes_arr[-1]:
+                warnings.warn(
+                    f"Fitted MDE ({mde:.4f}) is outside the evaluated range "
+                    f"[{effect_sizes_arr[0]:.4f}, {effect_sizes_arr[-1]:.4f}]. "
+                    f"Consider widening the effect_sizes range.",
+                    stacklevel=2,
+                )
+        except RuntimeError as e:
```
Comment (daimon-pymclabs) [inline causalpy/checks/power_analysis.py]: `except RuntimeError` is too narrow. `scipy.optimize.curve_fit` raises `ValueError` on degenerate inputs (NaNs, inconsistent bounds/p0) and can emit `OptimizeWarning` while returning a junk fit rather than raising at all. The `test_sigmoid_fitting_failure_warns` case happens to land on a RuntimeError, but flat/constant detection rates can surface as `ValueError`. Suggest `except (RuntimeError, ValueError)`.

Related: `LogisticFit.mde` computes `(1.0 / self.k) * ...` with `k` bounded `[0, inf]`. If the fit returns `k == 0`, this yields `inf` (numpy) or raises (python float); either way `mde` is meaningless. Consider guarding `k <= 0` and emitting the same 'fitting failed' warning path.

File: causalpy/tests/test_power_analysis.py [hunk 12]
```diff
@@ -0,0 +1,600 @@
...
+
+        assert len(grid_result.effect_sizes) == 20
+        assert len(sigmoid_result.effect_sizes) == 5
+        # Sigmoid should still produce a smooth curve with 200 points
+        assert len(sigmoid_result.smooth_effect_sizes) == 200
+
+
+# ---------------------------------------------------------------------------
+# Tests for error handling
+# ---------------------------------------------------------------------------
```
Comment (juanitorduz) [inline causalpy/tests/test_power_analysis.py]: and this one

File: causalpy/tests/test_power_analysis.py [hunk 13]
```diff
@@ -0,0 +1,600 @@
...
+        )
+        fig, ax = plt.subplots()
+        returned_fig = result.plot(ax=ax)
+        assert returned_fig is fig
+        plt.close(fig)
+
+
+# ---------------------------------------------------------------------------
+# Tests for coverage of edge branches
+# ---------------------------------------------------------------------------
```
Comment (juanitorduz) [inline causalpy/tests/test_power_analysis.py]: this

File: causalpy/checks/power_analysis.py [hunk 14]
```diff
@@ -0,0 +1,544 @@
...
+  Achieves ~40–60% computation reduction while producing smoother curves.
+
+The power analysis operates on the learned null distribution from a
+completed ``PlaceboInTime`` check.  It simulates what the design would
+conclude at each hypothetical true effect size using the same ROPE-based
+Bayesian decision rule.
+
+References
+----------
+Issue: https://github.com/pymc-labs/CausalPy/issues/820
```
Comment (juanitorduz) [inline causalpy/checks/power_analysis.py]: I think we can remove this issue reference

File: causalpy/checks/power_analysis.py [hunk 15]
```diff
@@ -0,0 +1,544 @@
...
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+"""
+Power analysis for geo-experiment and quasi-experiment designs.
+
+Provides efficient power curve estimation via two strategies:
```
Comment (juanitorduz) [inline causalpy/checks/power_analysis.py]: Do you have any references from the literature? (blogs or books?)

File: causalpy/checks/power_analysis.py [hunk 16]
```diff
@@ -0,0 +1,544 @@
...
+from causalpy.checks.base import CheckResult
+
+
+@dataclass
+class LogisticFit:
+    """Parameters of the fitted two-parameter logistic curve.
+
+    The model is::
+
+        P(detect | x) = 1 / (1 + exp(-k * (x - x0)))
```
Comment (juanitorduz) [inline causalpy/checks/power_analysis.py]: Could we wrap the math as in https://github.com/pymc-labs/CausalPy/blob/main/causalpy/experiments/synthetic_difference_in_differences.py#L76-L77 ?

File: causalpy/checks/power_analysis.py [hunk 17]
```diff
@@ -0,0 +1,544 @@
...
+    threshold: float
+    fitted_curve: LogisticFit | None = None
+    smooth_effect_sizes: np.ndarray | None = field(default=None, repr=False)
+    smooth_detection_rates: np.ndarray | None = field(default=None, repr=False)
+    mde: float | None = None
+
+    def plot(  # pragma: no cover
+        self,
+        power_threshold: float = 0.80,
+        ax: Any = None,
```
Comment (juanitorduz) [inline causalpy/checks/power_analysis.py]: Can we have better type hints? Something like pllt.Axes | None = None

File: causalpy/checks/power_analysis.py [hunk 18]
```diff
@@ -0,0 +1,544 @@
...
+        title: str = "Power Curve",
+        xlabel: str = "Effect size",
+        ylabel: str = "Detection probability",
+        show_mde: bool = True,
+    ) -> Any:
```
Comment (juanitorduz) [inline causalpy/checks/power_analysis.py]: better return type

File: causalpy/checks/power_analysis.py [hunk 19]
```diff
@@ -0,0 +1,544 @@
...
+    pit_result: CheckResult,
+    effect_sizes: list[float] | np.ndarray | None = None,
+    n_simulations: int = 200,
+    strategy: Literal["grid", "sigmoid"] = "grid",
+    n_evaluation_points: int = 5,
+    power_threshold: float = 0.80,
+    n_posterior_samples: int = 1000,
+    random_seed: int | None = None,
+) -> PowerCurveResult:
+    """Compute a power curve from a completed PlaceboInTime check.
```
Comment (juanitorduz) [inline causalpy/checks/power_analysis.py:429]: Can you please expand more on the description of the methodology?

File: causalpy/checks/power_analysis.py [hunk 20]
```diff
@@ -0,0 +1,544 @@
...
+        )
+        # Apply ROPE decision
+        prob_positive = float((simulated_posterior > rope_half_width).mean())
+        if prob_positive >= threshold:
+            detections += 1
+
+    return detections / n_simulations
+
+
+def power_analysis(
```
Comment (juanitorduz) [inline causalpy/checks/power_analysis.py:419]: This function is too long and contains many things... could we please modularize into smaller functions so that it's easier to read and unit-test :)

**Review by juanitorduz — CHANGES_REQUESTED** [review]
Hi @genrichez ! Thank you vary much for your contribution. I left some innitial comments on the code. I will take a detailed look into the logic next week :)

**Comment by juanitorduz** [comment]
by the way daimon-pymclabs in an internal agent reviewer ;)

File: causalpy/tests/test_power_analysis.py (hunk 1 above)
Comment (genrichez) [reply causalpy/tests/test_power_analysis.py]: Removed

File: causalpy/tests/test_power_analysis.py (hunk 2 above)
Comment (genrichez) [reply causalpy/tests/test_power_analysis.py]: Removed

File: causalpy/tests/test_power_analysis.py (hunk 4 above)
Comment (genrichez) [reply causalpy/tests/test_power_analysis.py]: And this one as well :)

File: causalpy/tests/test_power_analysis.py (hunk 5 above)
Comment (genrichez) [reply causalpy/tests/test_power_analysis.py]: Removed

File: causalpy/tests/test_power_analysis.py (hunk 6 above)
Comment (genrichez) [reply causalpy/tests/test_power_analysis.py]: Removed

File: causalpy/tests/test_power_analysis.py (hunk 12 above)
Comment (genrichez) [reply causalpy/tests/test_power_analysis.py]: Removed

File: causalpy/tests/test_power_analysis.py (hunk 13 above)
Comment (genrichez) [reply causalpy/tests/test_power_analysis.py]: Removed

File: causalpy/checks/power_analysis.py (hunk 14 above)
Comment (genrichez) [reply causalpy/checks/power_analysis.py]: Removed

File: causalpy/checks/power_analysis.py (hunk 15 above)
Comment (genrichez) [reply causalpy/checks/power_analysis.py]: Added , http://www.stat.columbia.edu/~gelman/research/published/retropower_final.pdf + https://psycnet.apa.org/doiLanding?doi=10.1037/a0029146

File: causalpy/checks/power_analysis.py (hunk 16 above)
Comment (genrichez) [reply causalpy/checks/power_analysis.py]: Done

File: causalpy/checks/power_analysis.py (hunk 18 above)
Comment (genrichez) [reply causalpy/checks/power_analysis.py]: Done

File: causalpy/checks/power_analysis.py (hunk 19 above)
Comment (genrichez) [reply causalpy/checks/power_analysis.py:429]: Elaborated on that

File: causalpy/checks/power_analysis.py (hunk 20 above)
Comment (genrichez) [reply causalpy/checks/power_analysis.py:419]: Modularized it, please double check it, thanks!

**Comment by genrichez** [comment]
> by the way daimon-pymclabs in an internal agent reviewer ;)

Got it :) 
Thanks for the feedback , implemented the changes, could you please re-review, thanks!

**Review by daimon-pymclabs — COMMENTED** [review]
Re-review after the latest push (`a600dfd`). Thanks @genrichez — the split into `_simulate_detection_rate` / `_extract_null_distribution` / `_build_effect_sizes` / `_fit_sigmoid` reads much better. Two of my earlier points are confirmed resolved:

- ✅ **Null draw** is now `rng.choice(null_samples)` (was the deterministic first-`n` prefix walk) — properly seeded and it samples the whole null distribution rather than a chain-slow prefix.
- ✅ **`_fit_sigmoid`** now catches `(RuntimeError, ValueError)`, so degenerate `curve_fit` inputs (NaNs, inconsistent bounds/`p0`) no longer escape.

One thing still outstanding, plus two optional nits:

1. **Docstring drift (please fix).** The `effect_sizes` docstring still says the default is `np.linspace(0, 4 * tau, 8)` (grid) / `[0, 4 * tau]` (sigmoid), but `_build_effect_sizes` uses `max(4.0 * tau, 3.0 * rope_half_width)` as the upper bound in both branches. Worth syncing the docstring so users know the ROPE half-width participates in the default range.
2. *(optional)* The ROPE decision is still inlined as `(simulated_posterior > rope_half_width).mean() >= threshold` rather than reusing `PlaceboInTime.bayesian_rope_decision`. Numerically equivalent to the one-sided `"positive"` branch today; only matters if the shared rule ever diverges.
3. *(optional)* The `fold_sds` → `fold_results`/`fr.fold_sd` fallback in `_extract_null_distribution` is still effectively unreachable against real `PlaceboInTime` output — a result carrying `null_samples` always carries `fold_sds` too. Harmless as defensive code.

Nothing blocking from my side beyond the docstring. Deferring to @juanitorduz on the detailed logic pass.

**Comment by drbenvincent** [comment]
## Automated triage

**Recommendation:** `review:high` — no decision gate identified.

**Why:**
- New public API: `causalpy.checks.power_analysis()` function, `PowerCurveResult` dataclass, and `LogisticFit` class.
- New module `causalpy/checks/power_analysis.py` (545 lines) with grid and sigmoid strategies.
- 556 lines of tests; all CI checks pass; PR is blocked (merge conflict).
- Review already has requested changes — on-going review.

**Review focus:**
1. Verify the sigmoid fitting strategy is numerically robust (handles flat data, degenerate ROPE, edge cases).
2. Confirm the public API integrates cleanly with the existing checks framework.
3. Check that the `PlaceboInTime` dependency assumptions are documented.

**Confidence:** high

**Comment by cetagostini** [comment]
Hey @genrichez! First off — this PR nails the important part: measuring detection against the **placebo-calibrated null** instead of a naive zero-null. That's the piece #819/#820 never got right, and the API you built around it (function + result object + plot + solid tests) is the right shape. 🙌

While validating it against the placebo-in-time notebook we stumbled on something fun though: the Monte Carlo is simulating a quantity that has an exact answer. The simulated posterior is `Normal(θ + e, σ)`, so the ROPE rule fires deterministically — `θ + e > rope + z·σ`, with `z = Φ⁻¹(threshold)`. The whole curve collapses to a boolean mean over the null draws × fold SDs, and the MDE is a one-liner:

```python
np.quantile(rope + z * sigma - theta, 0.80)
```

We ran both engines on the notebook's UK CO₂ fits and they agree within MC noise — but the sigmoid's MDE moves ±10% depending on the seed, while the exact one is deterministic and costs milliseconds. Once the exact curve is cheaper than *both* strategies, the sigmoid (and `n_simulations`, `n_posterior_samples`, `random_seed`, `strategy`...) doesn't have a job anymore. It also turns out the MC kernel here is a twin of the one inside `PlaceboInTime._compute_assurance` — same draws, same rule — so this is a chance to have one implementation instead of two.

So here's what I'd love to do, keeping your chassis exactly as-is:

- Swap the engine for the closed form — same API shape, but exact `p_detect` / `p_null` / `p_indeterminate` / `p_wrong_sign` curves, and MDE as the order statistic above. Sigmoid machinery and all the simulation knobs go away.
- Let `rope_half_width` / `threshold` be passed at analysis time — the notebook actually runs the check *without* a rope and picks it when plotting, which currently raises here.
- Rename to the vocabulary the library already uses: `operating_characteristics()` → `OperatingCharacteristics`, with `.fpr` and `.assurance(prior)`. "Power" is the frequentist word for it; our word for the prior-weighted scalar is already *assurance*, and the notebook already calls the curve *operating characteristics*.
- Point `_compute_assurance` at the same code path. (Only visible change: `AssuranceResult.null_decisions`/`alt_decisions` become `None` — the exact version has no per-replication decisions, and nothing in the repo used them.)

**I'm happy to give you a hand with this** — I actually have it working locally on top of your branch, tests and notebook included, plus an equivalence script showing it reproduces your MC numbers exactly. If you like, then I can push it as follow-up commits, so the PR stays yours; if you'd rather do it yourself, I'll share the harness and numbers and get out of your way 😄

Why I think it's worth it: everything becomes one piece of math instead of three. The curve at zero effect *is* the false-positive rate `PlaceboInTime` reports, assurance *is* the curve integrated against the expected-effect prior (we checked both identities numerically), and the notebook can call the library instead of carrying its own private version. Reader goes from notebook to API without switching vocabulary — and your PR ends up being the thing that unified it.

P.S. This would also let us park #820 — the sigmoid existed to save MC evaluations, and there's nothing left to save.

**Comment by cetagostini** [comment]
On the other hand, I can say "OperatingCharacteristics" it's the best name I could found to join different literature, and ideology, not super attach to it. But definitely not "Power" 😅

**Comment by cetagostini** [comment]
Hey @genrichez let me know! 🙌🏻
